### PR TITLE
 [bugfix][ogr] Recompute capabilities when subsetfilter is set

### PR DESCRIPTION
**Pull request by elpaso**
Fixes #17795 , with tests.

### CODE REVIEW

**Comment by elpaso** [comment]
@rouault please have a look

File: src/providers/ogr/qgsogrprovider.cpp [hunk 1]
```diff
@@ -4038,6 +4050,13 @@ void QgsOgrProvider::open( OpenMode mode )
     }
   }
 
+  // Connect
+  if ( mode == OpenModeInitial )
```
Comment (rouault) [inline src/providers/ogr/qgsogrprovider.cpp]: code indentation with the lambda is confusing. Perhaps wrap the connect() itself in { }

File: src/providers/ogr/qgsogrprovider.cpp [hunk 2]
```diff
@@ -607,7 +607,7 @@ bool QgsOgrProvider::setSubsetString( const QString &theSQL, bool updateFeatureC
   invalidateCachedExtent( false );
 
   // Changing the filter may change capabilities
-  computeCapabilities();
+  emit capabilitiesNeedUpdate();
```
Comment (rouault) [inline src/providers/ogr/qgsogrprovider.cpp]: why to make a signal for that ? Nothing outside of the provider is interested in that. Or does it cause some asynchronous call ?

Comment (elpaso) [reply src/providers/ogr/qgsogrprovider.cpp]: yes, that was the idea: we don't want the update capabilities function to be called on first call of setSubsetString but only for following changes that would occour to the filter. If you prefer, we could add an argument to setSubsetString to tell wether the capabilities needs to be updated or not.
But maybe we could just call computeCapabilities all the times and remove the explicit call to computeCapabilities from the open() code, the effect would be the same.

Comment (rouault) [reply src/providers/ogr/qgsogrprovider.cpp]: > we could add an argument to setSubsetString to tell wether the capabilities needs to be updated or not.

Would seem more obvious to me. signals are hard to follow

Comment (elpaso) [reply src/providers/ogr/qgsogrprovider.cpp]: mmm, setSubsetString is an override, we would need to change all providers.  I actually added an asynchronous call to avoid recursion when setSubSetString is invoked from open(), the update would call open() again and again for shapefiles.
I think the signal is our only chance in this case (I'd really prefer to not change all providers API for such a particular case).

Comment (rouault) [reply src/providers/ogr/qgsogrprovider.cpp]: oh well, so be it. some code comments to capture this will help later readers of that code :-)

Comment (elpaso) [reply src/providers/ogr/qgsogrprovider.cpp]: @rouault I spoke too early, see my last commit for a connect-less implementation (easier to backport too).
What do you think?

Comment (rouault) [reply src/providers/ogr/qgsogrprovider.cpp]: > What do you think?

+1
